presto-main/src/main/java/com/facebook/presto/execution/ThreadWatchdog.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.execution;
+
+import io.airlift.units.Duration;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * ThreadWatchdog spins up a background task
+ * that interrupts given thread after the timeout expires.
+ */
+public class ThreadWatchdog
+        implements AutoCloseable
+{
+    private final Thread thread;
+    private final ScheduledExecutorService executor;
+    private final Duration timeout;
+
+    private Future<?> future;
+
+    /**
+     * @param thread Thread to be interrupted after timeout expires.
+     * @param executor Executor to run the background watchdog task.
+     * @param timeout Expiration time of the thread.
+     */
+    public ThreadWatchdog(Thread thread, ScheduledExecutorService executor, Duration timeout)
+    {
+        this.thread = requireNonNull(thread);
+        this.executor = requireNonNull(executor);
+        this.timeout = requireNonNull(timeout);
+
+        this.start();
+    }
+
+    private void start()
+    {
+        this.future = executor.schedule(
+                thread::interrupt, timeout.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
In j.u.c usage you normally always do nanoseconds to allow the highest precision timings to flow through. It doesn't matter here though.